# Data Cleaning

In [1]:
# import relevant libraries
import pandas as pd

In [62]:
# import all raw data into datasets

df_customers = pd.read_csv("data/raw/olist_customers_dataset.csv")
df_geolocation = pd.read_csv("data/raw/olist_geolocation_dataset.csv")
df_order_items = pd.read_csv("data/raw/olist_order_items_dataset.csv")
df_order_payments = pd.read_csv("data/raw/olist_order_payments_dataset.csv")
df_order_reviews = pd.read_csv("data/raw/olist_order_reviews_dataset.csv")
df_orders = pd.read_csv("data/raw/olist_orders_dataset.csv")
df_products = pd.read_csv("data/raw/olist_products_dataset.csv")
df_sellers = pd.read_csv("data/raw/olist_sellers_dataset.csv")
df_customers

customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

       customer_zip_code_prefix          customer_city customer_state  
0                         14409                 franca             SP  
1                          9790  sao bernardo do campo             SP  
2                          1151              sao paulo             SP  
3                          8775        mogi das cruzes             SP  
4                         13056               campinas             SP  
...                         ...                    ...            ...  
99436                      3937              sao paulo             SP  
99437                      6764        taboao da serra             SP  
99438                     60115              fortaleza             CE  
99439                     92120                 canoas             RS  
99440                      6703                  cotia             SP  

[99441 rows x 5 columns]

## Customers


In [7]:
# having a look at the data
df_customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [9]:
# checking data types
df_customers.dtypes

customer_id                 object
customer_unique_id          object
customer_zip_code_prefix     int64
customer_city               object
customer_state              object
dtype: object

In [22]:
# looking to see which is a primary key
df_customers["customer_unique_id"].nunique()

96096

In [23]:
# looking to see which is a primary key
df_customers["customer_id"].nunique()

99441

It appears that the primary key is the customer_id and some customers appear in the data twice

In [44]:
# dropping duplicate rows
df_customers.dropna(inplace=True)
df_customers.shape

(99441, 5)

No duplicate rows found

In [15]:
# looking for null values
df_customers.isnull().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [30]:
# Looking at number of distinct values per column
print(f"There are {df_customers['customer_state'].nunique()} different states.")
print(f"There are {df_customers['customer_city'].nunique()} different cities.")
print(f"There are {df_customers['customer_zip_code_prefix'].nunique()} different ZIP code prefixes.")

There are 27 different states.
There are 4119 different cities.
There are 14994 different ZIP code prefixes.


The data is clean and can be sent directly to be enriched

In [31]:
# exporting to the clean folder
df_customers.to_csv("data/cleaned/clean_customers.csv", index=False)

## Geolocation


In [32]:
# quick glance at the data
df_geolocation.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

In [35]:
# checking data types
df_geolocation.dtypes

geolocation_zip_code_prefix      int64
geolocation_lat                float64
geolocation_lng                float64
geolocation_city                object
geolocation_state               object
dtype: object

In [38]:
# checking the shape
df_geolocation.shape

(1000163, 5)

In [37]:
# checking for a primary key
df_geolocation["geolocation_zip_code_prefix"].nunique()

19015

This isn't a dataset of zip codes with their relevant rough geolcations, but rather a massive dump if lat/lon geolocations that have their relevant zip/city/state. This dataset is probably the least useful as this level of granularity is probably overkill

In [45]:
# dropping duplicate rows
df_geolocation.dropna(inplace=True)
df_geolocation.shape

(1000163, 5)

No duplicate rows found

In [40]:
# looking for null values
df_geolocation.isnull().sum()

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

In [41]:
# Checking to see if the columns match the customer columns or exceed it
print(f"There are {df_geolocation['geolocation_state'].nunique()} different states.")
print(f"There are {df_geolocation['geolocation_city'].nunique()} different cities.")
print(f"There are {df_geolocation['geolocation_zip_code_prefix'].nunique()} different ZIP code prefixes.")

There are 27 different states.
There are 8011 different cities.
There are 19015 different ZIP code prefixes.


This could be that there are sellers who are at locations that there are no customers, or that this dump contains all locations within a range whether there is a seller or customer there. Or both. But it suggests there are likely no missing locations if a join with customers is necessary

In [54]:
import folium
from IPython.display import display

# Define the minimum and maximum latitude and longitude coordinates
min_lat, max_lat = df_geolocation["geolocation_lat"].min(), df_geolocation["geolocation_lat"].max()
min_lon, max_lon = df_geolocation["geolocation_lng"].min(), df_geolocation["geolocation_lng"].max()

# Calculate the center of the rectangle
center_lat = (min_lat + max_lat) / 2
center_lon = (min_lon + max_lon) / 2

# Create a map centered around the center of the rectangle
mymap = folium.Map(location=[center_lat, center_lon], zoom_start=10)

# Add a rectangle to the map
folium.Rectangle(
    bounds=[[min_lat, min_lon], [max_lat, max_lon]],
    fill=True,
    fill_color='blue',
    fill_opacity=0.2,
    color='blue',
    opacity=0.4,
    weight=2
).add_to(mymap)

# Display map
display(mymap)

This shows the maximum area covered in the geo data. Note the entirety of Brazil and Africa but NOT USA, Australia, Asia or Europe. Especial note to states like California and texas not being part of the data, gives a clue as to the customer/sellers base

The data is clean and can be sent directly to be enriched

In [46]:
# exporting to the clean folder
df_geolocation.to_csv("data/cleaned/clean_geolocation.csv", index=False)

## Order Items

In [47]:
# quick glance at the data
df_order_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [49]:
# checking data types
df_order_items.dtypes

order_id                object
order_item_id            int64
product_id              object
seller_id               object
shipping_limit_date     object
price                  float64
freight_value          float64
dtype: object

In [50]:
#  converting the shipping_limit_date to a datetime format
df_order_items["shipping_limit_date"] = pd.to_datetime(df_order_items["shipping_limit_date"])
df_order_items.dtypes

order_id                       object
order_item_id                   int64
product_id                     object
seller_id                      object
shipping_limit_date    datetime64[ns]
price                         float64
freight_value                 float64
dtype: object

In [56]:
df_order_items.describe()

order_item_id            shipping_limit_date          price  \
count  112650.000000                         112650  112650.000000   
mean        1.197834  2018-01-07 15:36:52.192685312     120.653739   
min         1.000000            2016-09-19 00:15:34       0.850000   
25%         1.000000     2017-09-20 20:57:27.500000      39.900000   
50%         1.000000            2018-01-26 13:59:35      74.990000   
75%         1.000000  2018-05-10 14:34:00.750000128     134.900000   
max        21.000000            2020-04-09 22:35:08    6735.000000   
std         0.705124                            NaN     183.633928   

       freight_value  
count  112650.000000  
mean       19.990320  
min         0.000000  
25%        13.080000  
50%        16.260000  
75%        21.150000  
max       409.680000  
std        15.806405

Average price of 120, with min of 0.85 and max of 6,735. Doesn't raise any red flags immediately.

Average freight value of 20, with min of 0 and max of 409. Doesn't raise any red flags neither. 

Date ranges from Sep 2016 to April 2020 for 3.5 years of data

In [57]:
# checking the shape
df_order_items.shape

(112650, 7)

In [60]:
# checking for a primary key
print(f"{df_order_items['order_id'].nunique()} unique values for order_id.")
print(f"{df_order_items['order_item_id'].nunique()} unique values for order_item_id.")
print(f"{df_order_items['product_id'].nunique()} unique values for product_id.")
print(f"{df_order_items['seller_id'].nunique()} unique values for seller_id.")

98666 unique values for order_id.
21 unique values for order_item_id.
32951 unique values for product_id.
3095 unique values for seller_id.


None of these are a primary key, suggesting this is the list of products per order. So the primary key would be the combination of the order_id and the order_item_id (each individual product within the order). It wouldn't be order_id and product_id as there are likely orders that contain the same product twice and since there is no quantity value here the way to represent that would be to just list the product id and order id for each ordered.

In [67]:
# given this, rather than dropping the duplicate rows its best to check if any exist:
df_order_items[df_order_items.duplicated(keep=False)]
df_order_items.shape

(112650, 7)

Better to check before removing sales data, but it seems to be as I suspected

In [69]:
# looking for null values
df_order_items.isnull().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

The data is clean and can be sent directly to be enriched

In [70]:
# exporting to the clean folder
df_order_items.to_csv("data/cleaned/clean_order_items.csv", index=False)

## Order Payments

In [71]:
## Order Payments

In [72]:
# quick glance at the data
df_order_payments.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45